Start deep dive docs (#11059)
conda/conda: docs/source/dev-guide/deep-dive-install.md

optimization step also takes longer the bigger the index gets.
```

### Channel priorities

`context.channels` returns an `IndexedSet` of `Channel` objects; essentially a list of unique
items. The different channels in this list can have overlapping or even conflicting information
for the same package name. For example, `defaults` and `conda-forge` will for sure contain
packages that fullfil the `conda install numpy` request. Which one is chosen by `conda` in this
case? It depends on the `context.channel_priority` setting: From the help message:

> Accepts values of 'strict', 'flexible', and 'disabled'. The default value is 'flexible'. With
> strict channel priority, packages in lower priority channels are not considered if a package
> with the same name appears in a higher priority channel. With flexible channel priority, the
> solver may reach into lower priority channels to fulfill dependencies, rather than raising an
> unsatisfiable error. With channel priority disabled, package version takes precedence, and
> the configured priority of channels is used only to break ties.

In practice, `channel_priority=strict` is often the recommended setting for most users. It's faster

> to solve and causes less problems down the line. Check more details

Reviewer: "fewer problems"

<!--
Author: Jaime Rodríguez-Guerra <@jaimergp>, Quansight Labs.

Some resources that helped draft this document:
* https://speakerdeck.com/asmeurer/conda-internals
* https://www.anaconda.com/blog/understanding-and-improving-condas-performance
-->

(deep_dive_install)=

# Deep dive: conda install

In this document we will explore what happens in Conda from the moment a user types their
installation command until the process is finished successfully. For the sake of completeness,
we will consider the following situation:

* The user is running commands on a Linux x64 machine with a working installation of Miniconda.
* This means we have a `base` environment with `conda`, `python`, and their dependencies.
* The `base` environment is already preactivated for Bash. For more details on activation, check
  {ref}`deep_dive_activation`.

So... here we are:

```console
$ (base) ~/: conda info
... # TODO

$ (base) ~/: conda list
... # TODO

> ```

Reviewer 2: I think this is missing some info.

in `~/.bashrc` and others. If you are curious where his happens, it's
[`conda.activate`][conda.activate].

Since our command is `conda install ...`, we still need to arrive somewhere else. You will notice
that the rest of the logic is delegated to `conda.cli.main:_main()`, which will invoke the parser
generators, initialize the context and loggers, and, eventually, pass the argument list over to
the corresponding command function. These four steps are implemented in four functions/classes:

1. [`conda.cli.conda_argparse:generate_parser()`][conda.cli.conda_argparse:generate_parser]:
  This uses `argparse` to generate the CLI. Each subcommand is initialized in separate functions.
  Note that the command-line options are not generated dynamically from the `Context` object, but
  annotated manually. If this is needed (e.g. `--repodata-fn` is exposed in `Context.repodata_fn`),
  the `dest` variable of each CLI option should
  [match the target attribute in the context object][context_match].
2. [`conda.base.context.Context`][context_init]:
  Initialized taking into account, among other things, the parsed arguments from step above. This is

>   covered in more detail in a separate deep dive: {ref}`deep_dive_context`.

Author: Added a few more words here to clarify the message!

>   covered in more detail in a separate deep dive: {ref}`deep_dive_context`.

Reviewer 3: The rewording looks good!

  metadata for each package available on that platform.
* In most cases, the same subdirs also contain the `*.tar.bz2` files for each of the published
  packages. This is what `conda` downloads and extracts once solving is complete. The anatomy
  of these files is well defined, both in content and naming structure. See
  {ref}`concept-conda-package`, {ref}`package_metadata` and/or [Package naming
  conventions][conda_build_package_names] for more details.

Additionally, the channel main directory might contain a `channeldata.json` file, with channel-wide
metadata (this is, not specific per platform). Not all channels include this and, actually, it's not
really used these days. <!-- TODO: Check why and reasons for this file to exist -->

Since conda's philosophy is to keep all packages ever published around for reproducibility,
`repodata.json` is always growing, which presents a problem both for the download itself and the
solver engine. To reduce download times and bandwidth usage, `repodata.json` is also served as a
BZIP2 compressed file, `repodata.json.bz2`. This is what most `conda` clients end up downloading.

> ```{admonition} Note on current_repodata.json

Author: I am not entirely sure this renders correctly. Need to check.

prefix location. Instead, `conda` maintains a cache of every package ever downloaded to disk (both
the tarball and the extracted contents). To save space and speed up environment creation and
deletion, files are not copied over, but instead they are linked (usually via a hardlink). That's
why these two tasks are separated in the transaction logic: you don't need to download and extract
packages that are already in the cache; you only need to link them!

```{admonition} Transactions also drive reports

The type and number of actions can also be calculated by `_make_legacy_action_groups()`, which
returns a list of _action groups_ (one per `PrefixSetup`). Each action group is a just a dictionary
following this specification:

  ```
  {
    "FETCH": Iterable[PackageRecord],  # estimated by `ProgressiveFetchExtract`
    "PREFIX": str,
    "UNLINK": Iterable[PackageRecord],
    "LINK: Iterable[PackageRecord],
  }
  ```

These simpler action groups are only used for reporting, either via a processed text report
(via `print_transaction_summary`) or just the raw JSON (via `stdout_json_success`). As you can see,
they do not know anything about other types of tasks.

> ```

Reviewer 3: The formatting here is wonky, please see screenshot below for reference:

![Screen Shot 2022-02-14 at 9 58 43 PM](https://user-images.githubusercontent.com/28930622/153984384-0041aab8-4324-4033-9fed-61e9ca005cef.png)

It might not be possible to have a code snippet in the middle of an admonition block, unfortunately.

Author: Oh, I think I was missing a set of backticks in the outer level.

Author: Can you check with the last changes in b906271?

Reviewer 3: Looks like you fixed it! 👍 

![Screen Shot 2022-02-15 at 9 48 29 AM](https://user-images.githubusercontent.com/28930622/154085975-59f99e45-09d4-4c58-99be-6996d82afac8.png)

## Download and extraction

`conda` maintains a cache of downloaded tarballs and their extracted contents to save disk space
and improve the performance of environment modifications. This requires some code to check whether
a given `PackageRecord` is already present in the cache, and, if it's not, how to download the
tarball and extract its contents in a performant way. This is all handled by the
`ProgressiveFetchExtract` class, which can instantiate up to two `Action` objects for each
passed `PackageRecord`:

* `CacheUrlAction`: downloads (if remote) or copies (if local) a tarball to the cache location.
* `ExtractPackageAction`: extracts the contents of the tarball.

These two actions only take place _if_ the package is not in cache yet and if it has already been
extracted, respectively. They can also revert the changes if the transaction is aborted (either
due to an error or because the user pressed Ctrl+C).

```{admonition} Anatomy of the conda caches
TODO

> ```

Reviewer 2: Also a TODO?

Author: I commented this out for now, not crucial to understanding the whole thing.

transaction execution logic expects. To support all the different actions involved in populating
the prefix, the `PathAction` class tree holds quite the graph:

```
PathAction
  PrefixPathAction
    CreateInPrefixPathAction
      LinkPathAction
        PrefixReplaceLinkAction
      MakeMenuAction
      CreateNonadminAction
      CreatePythonEntryPointAction
      CreatePrefixRecordAction
      UpdateHistoryAction
    RemoveFromPrefixPathAction
      UnlinkPathAction
        RemoveLinkedPackageRecordAction
      RemoveMenuAction
  RegisterEnvironmentLocationAction
  UnregisterEnvironmentLocationAction
  CacheUrlAction
  ExtractPackageAction

MultiPathAction
  CompileMultiPycAction
    AggregateCompileMultiPycAction

```

You are welcome to read on the docstring for each of those classes to understand which each one
is doing; all of them are listed under `conda.core.path_actions`. In the following sections, we will
only comment on the most important ones.

## Linking the files in the environment

When conda _links_ a file from the cache location to the prefix location, it can actually mean
three different actions:

1. Creating a soft link
2. Creating a hard link
3. Copying the file

> The difference between softlinks and hardlinks is subtle, but important. You can find more info on
> the differences elsewhere (e.g. here), but for our purposes it means that:

Reviewer 3: ```suggestion
The difference between soft links and hard links is subtle, but important. You can find more info on
the differences elsewhere (e.g. here), but for our purposes it means that:
```

Additionally, the "(e.g. here)" bit seems to be missing a link

* Hard links are cheaper to resolve, behave like a real file, but can only link files in the same
  mount point.
* Soft links can link files across mount points, but they don't behave exactly like files (more like
  forwarders), so it's possible that they break assumptions made in certain pieces of code.

> Most of the time, `conda` will try to hardlink files and, if that fails, it will copy them over.

Reviewer 3: ```suggestion
Most of the time, `conda` will try to hard link files and, if that fails, it will copy them over.
```

Copying a file is an expensive disk operation, both in terms of time and space, so it should be
the last option. However, sometimes it's the only way. Especially, when the file needs to be
modified to be used in the target prefix.

Ummm... what? Why would `conda` modify a file to install it? This has to do with relocatability.
When a `conda` package is created, `conda-build` creates up to three temporary environments:

> * Build environment: where compilers and other build tools are installed to, separate from the
>   host environment to support cross-compilation.
> * Host environment: where build-time dependencies are installed to, together with the package you
>   are building.
> * Test environment: where run-time dependencies are installed, together with the package you just
>   build. It simulates what will happen when a user installs the package so you can run arbitrary
>   checks on your package.

Reviewer 3: ```suggestion
* Build environment: where compilers and other build tools are installed, separate from the
  host environment to support cross-compilation.
* Host environment: where build-time dependencies are installed, together with the package you
  are building.
* Test environment: where run-time dependencies are installed, together with the package you just
  built. It simulates what will happen when a user installs the package so you can run arbitrary
  checks on your package.
```

never true for `conda` packages. They are created in one machine and installed in another. To avoid
"path not found" issues and other problems, `conda-build` marks those packages that hold references
to the build-time paths by replacing them with placeholders. At install-time, `conda` will replace
those placeholders with the target prefix and everything works!

But there's a problem: we can't modify the files on the cache location because they might be used
across environments (with obviously different paths). In these cases, files are not linked, but
copied; the path replacement only happens on the target copy, of course!

How does `conda` know how to link a given package or, more precisely, its extracted files? All
of this is determined in the preparation routines contained in
[`UnlinkLinkTransaction._prepare()`][conda.core.link:_prepare] (more specifically, through
[`determine_link_type()`][conda.core.link:determine_link_type]), as well as
[`LinkPathAction.create_file_link_actions()`][conda.core.path_actions:create_file_link_actions].

Note that the (un)linking actions also include the execution of pre-(un)link and post-(un)link
scripts, if listed.

> ## Action groups and actions, in detal

Reviewer 3: ```suggestion
## Action groups and actions, in detail
```

Once the old packages have been removed and the new ones have been linked through the appropriate

> means, we are done, right? Well, not there yet! There's one step left: the post-linking logic.

Reviewer 3: ```suggestion
means, we are done, right? Not yet! There's one step left: the post-linking logic.
```

3. Solving the install request
    * Requested packages + prefix state = list of specs
    * Index reduction (sometimes)
    * Running the solver
    * Post-processing the list of packages
4. Generating the transaction and the corresponding actions
5. Download and extraction
6. Integrity verification
7. Linking and unlinking files
8. Post-linking and post-activation tasks

## Command-line interface

<!-- This might be better in a separate deep-dive for the command line alone -->

First, a quick note on an implementation detail that might be not obvious.

When you type `conda install numpy` in your terminal, Bash took those three words and looked for a
`conda` command to pass a list of arguments `['conda', 'install', 'numpy']`. Before finding the
`conda` executable located at `CONDA_HOME/condabin`, it probably found the shell function defined
[here][conda_shell_function]. This shell function runs the (de)activation logic on the shell if
requested, or delegates over to the actual Python entry-points otherwise. This part of the logic can
be found in [`conda.shell`][conda.shell].

Once we are running the Python entry-point, we are in the [`conda.cli`][conda.cli]

> reigns. The function called by the entry point is [`conda.cli.main:main()`][conda.cli.main:main].

Author: Oops meant realm I guess.

[conda_cli_install_presolve_logic]: https://github.com/conda/conda/blob/4.11.0/conda/cli/install.py#L107
[conda_install_delegation]: https://github.com/conda/conda/blob/4.11.0/conda/cli/conda_argparse.py#L775
[conda_main_install_delegation]: https://github.com/conda/conda/blob/4.11.0/conda/cli/main_install.py#L12
[conda_shell_function]: https://github.com/conda/conda/blob/4.11.0/conda/shell/etc/profile.d/conda.sh#L62-L76
[conda.activate]: https://github.com/conda/conda/blob/4.11.0/conda/activate.py
[conda.cli.conda_argparse:do_call]: https://github.com/conda/conda/blob/4.11.0/conda/cli/conda_argparse.py#L77
[conda.cli.conda_argparse:generate_parser]: https://github.com/conda/conda/blob/4.11.0/conda/cli/conda_argparse.py#L28
[conda.cli.install:install]: https://github.com/conda/conda/blob/4.11.0/conda/cli/install.py#L107
[conda.cli.main_install]: https://github.com/conda/conda/blob/4.11.0/conda/cli/main_install.py
[conda.cli.main:main]: https://github.com/conda/conda/blob/4.11.0/conda/cli/main.py#L121
[conda.cli]: https://github.com/conda/conda/tree/4.11.0/conda/cli
[conda.core.index:get_reduced_index]: https://github.com/conda/conda/blob/4.11.0/conda/core/index.py#L246

> [conda.core.link:_prepare]: https://github.com/conda/conda/blob/4.11.0/conda/core/link.py#L266

Reviewer: You've got two of these "conda.core.link:_prepare" links with the same name, but different links.
